bundles/org.openhab.binding.updateopenhab/src/main/java/org/openhab/binding/updateopenhab/internal/UpdateHandler.java

```diff
@@ -0,0 +1,134 @@
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.updateopenhab.internal;
+
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.openhab.binding.updateopenhab.scripts.BaseUpdater;
+import org.openhab.binding.updateopenhab.scripts.DebianUpdater;
+import org.openhab.binding.updateopenhab.scripts.MacUpdater;
+import org.openhab.binding.updateopenhab.scripts.WindowsUpdater;
+import org.openhab.core.library.types.OnOffType;
+import org.openhab.core.thing.ChannelUID;
+import org.openhab.core.thing.Thing;
+import org.openhab.core.thing.ThingStatus;
+import org.openhab.core.thing.ThingStatusDetail;
+import org.openhab.core.thing.binding.BaseThingHandler;
+import org.openhab.core.types.Command;
+import org.openhab.core.types.RefreshType;
+import org.openhab.core.types.UnDefType;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link UpdateHandler} is responsible for handling commands, which are
+ * sent to one of the channels.
+ *
+ * @author Andrew Fiddian-Green - Initial contribution
+ */
+@NonNullByDefault
+public class UpdateHandler extends BaseThingHandler {
+
+    protected final Logger logger = LoggerFactory.getLogger(UpdateHandler.class);
+
+    private TargetVersion targetVersion;
+    private @Nullable BaseUpdater updater;
+    private @Nullable ScheduledFuture<?> refreshTask;
+
+    /**
+     * Constructor
+     */
+    public UpdateHandler(Thing thing) {
+        super(thing);
+        targetVersion = TargetVersion.STABLE;
+    }
+
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        if (command instanceof RefreshType) {
+            updateChanels();
+            return;
+        }
+        if (BindingConstants.CHANNEL_UPDATE_COMMAND.equals(channelUID.getId()) && (OnOffType.ON.equals(command))) {
+            scheduler.schedule(() -> {
+                updateState(channelUID, OnOffType.OFF);
+            }, 1, TimeUnit.SECONDS);
+            if (updater != null) {
+                scheduler.submit(updater);
+            }
+        }
+    }
+
+    @Override
+    public void initialize() {
+        Configuration config = getConfigAs(Configuration.class);
+        try {
+            targetVersion = TargetVersion.valueOf(config.targetVersion);
+        } catch (IllegalArgumentException e) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR);
+            return;
+        }
+
+        OperatingSystem os = OperatingSystem.getOperatingSystemVersion();
+        updateProperty(BindingConstants.PROPERTY_OPERATING_SYSTEM, os.toString());
+        switch (os) {
+            case MAC:
+                updater = new MacUpdater(targetVersion, config.password);
+                break;
+            case UNIX:
+                updater = new DebianUpdater(targetVersion, config.password);
```

Comment:
What happens if this is a non Debian based system like Redhat or Gentoo?

Reply:
Each version has a class of Updater that extends BaseUpdater. I wrote three so far — namely WindowsUpdater, DebianUpdater, and MacUpdater. The former two, I can test because I have such computers. The latter is basically a guess; half way between Windows and Debian, I think. Any other variety of system would need volunteers to write and test (hopefully largely copy/ paste/modify.

Reply:
PS I have also written an interface decoupler class that reads/writes between the xxUpdater class and the OH Thing handler, so one can change the binding, or the updater, fully independently.

Reply:
Cool,  you may want to check for a debian flavor specifically,  something like 

`Files.readString(Path.of("/preoc/version")).indexOf("Debian")  > 0 ` 

 so not to match non debian based OSs

Reply:
Hi @digitaldan many thanks for the suggestion. Just for info, it is not so much the OS distro that needs to be detected, but rather their respective Package Manager (which we use to run the actual update) -- be it debian (apt), rpm (yum, dnf), gentoo etc.  Anyway I think the better place to discover more appropriate the details is in `/etc/os-release` - any thoughts?

Reply:
>be it debian (apt), rpm (yum, dnf), gentoo etc.

I was just using Debian as an example here as thats what you have named the the updater class, but checking for package manager vs distro sounds like a good plan.

>yway I think the better place to discover more appropriate details is in /etc/os-release - any thoughts?

Sounds good to me, my suggestion for the version was just what i know off the top of my head :-)